node/src/Decoder.ts

```diff
@@ -0,0 +1,33 @@
+export interface Decoder<T> {
+    decode(input: string | Buffer): T;
+}
+
+class BytesDecoder implements Decoder<any> {
+    decode(input: string | Buffer): any {}
+}
+
+class StringDecoder implements Decoder<any> {
+    decode(input: any): any {
+        if (Buffer.isBuffer(input)) {
```

Comment:
I would try and do an inplace manipulation instead of returning (did not check if this is possible) - this will make your life easier along the way